src/Processors/Formats/ISchemaReader.cpp

```diff
@@ -223,8 +223,10 @@ NamesAndTypesList IRowWithNamesSchemaReader::readSchema()
             break;
 
         std::unordered_set<std::string_view> names_set; /// We should check for duplicate column names in current row
-        for (auto & [name, new_type] : new_names_and_types)
+        for (auto & new_name_and_type : new_names_and_types)
         {
+            auto & name = new_name_and_type.name;
+            auto & new_type = new_name_and_type.type;
             if (names_set.contains(name))
```

Comment:
Seems like whil using `for auto & [name, new_type] : new_names_and_types`
`name` and `new_type` are not references but copied objects.
Maybe because we have only specialization for `const const NameAndTypePair &` here:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/NamesAndTypes.h#L53-L62

Reply:
I tried to add specialization for `NameAndTypePair & name_and_type` but it didn't help for some reason...

Reply:
And we want to change the `new_type` in `chooseResultColumnType`, ok.

And why do we need such a big test case for this, why it can't be reproduced with smaller one?

Reply:
It was reproduced only on large files. But I just realized that with address sanitizer it will reproduce with 1 row. Let me update the test

Reply:
A Fuzzer of data formats will be helpful.

Reply:
>I tried to add specialization for NameAndTypePair & name_and_type but it didn't help for some reason...

To make it work you have to make sure that:
 1. Your return type is reference e.g. `return std::ref(name_and_type.name);` (or just avoid using `decltype(auto)`) and make sure that returned
 2. Tuple also holds reference, not value `template <> struct tuple_element<0, DB::NameAndTypePair> { using type = DB::String&; };`

Reply:
Thanks! Let me try it